#!/bin/bash
#PBS -l nodes=1:ppn=4
#PBS -N Post_Medusa_Genome_Processing
#PBS -l walltime=24:00:00
#PBS -j oe
#PBS -o ${HOME}/STD_OUT/${PBS_JOBNAME}_${PBSNAME}_${PBS_JOBID}.out

cd $PBS_O_WORKDIR

JOBSDIR=${BASEDIR}/jobScripts

if [[ ${DELLY} == "No" ]]
then

	# Make needed directories for Genome

	#{{{ 

	ssh ${USER}@${DATAMOVER} "mkdir -p /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/cna_manual_2016/${GENOMEPAIR}_unfi"

	if [ $? -ne 0 ]
	then
		rm CNA_Manual_In_Progress
		rm cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_In_Progress
		rm cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_In_Progress
		echo Failed to make cna_manual unfi dir on dback ${GENOMEPAIR} >> CNA_Manual_Fail
	        echo Failed to make cna_manual unfi dir on dback ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_Fail
		echo Failed to make cna_manual unfi dir on dback ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_Fail
	        echo Failed to make cna_manual unfi dir on dback ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_Fail
	        echo Failed to make cna_manual unfi dir on dback ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
	        exit 1
	fi

	ssh ${USER}@${DATAMOVER} "mkdir -p /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/cna_manual_2016/${GENOMEPAIR}_filt2012"

	if [ $? -ne 0 ]
	then
		rm CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_In_Progress
		rm cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_In_Progress
		echo Failed to make cna_manual filt dir on dback ${GENOMEPAIR} >> CNA_Manual_Fail
	        echo Failed to make cna_manual filt dir on dback ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_Fail
		echo Failed to make cna_manual filt dir on dback ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_Fail
	        echo Failed to make cna_manual filt dir on dback ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_Fail
	        echo Failed to make cna_manual filt dir on dback ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
	        exit 1
	fi

	ssh ${USER}@${DATAMOVER} "mkdir -p /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/cna_manual_2016/${GENOMEPAIR}_filt2016"

	if [ $? -ne 0 ]
	then
	        rm CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_In_Progress
		rm cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_In_Progress
	        echo Failed to make cna_manual filt dir on dback ${GENOMEPAIR} >> CNA_Manual_Fail
	        echo Failed to make cna_manual filt dir on dback ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_Fail
		echo Failed to make cna_manual filt dir on dback ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_Fail
	        echo Failed to make cna_manual filt dir on dback ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_Fail
	        echo Failed to make cna_manual filt dir on dback ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
	        exit 1
	fi

	#}}}

	# Rsync needed files over to dback for Unfiltered analysis

	#{{{

	# Genome Normal Dat
	ssh ${USER}@${DATAMOVER} "rsync ${STARTDIR}/${PATIENT_NAME}/${NASSAY}/${NORMALSAMPLEG}/${NORMALDAT} /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/cna_manual_2016/${GENOMEPAIR}_unfi/ "

	if [ $? -ne 0 ]
	then
		rm CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_In_Progress
		rm cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_In_Progress
		echo Failed to rsync Normal Dat file to dback for unfi ${GENOMEPAIR} >> CNA_Manual_Fail
		echo Failed to rsync Normal Dat file to dback for unfi ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_Fail
		echo Failed to rsync Normal Dat file to dback for unfi ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_Fail
		echo Failed to rsync Normal Dat file to dback for unfi ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_Fail
		echo Failed to rsync Normal Dat file to dback for unfi ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
		exit 1
	fi

	# Genome Tumor Dat

	ssh ${USER}@${DATAMOVER} "rsync ${STARTDIR}/${PATIENT_NAME}/${TASSAY}/${TUMORSAMPLEG}/${TUMORDAT} /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/cna_manual_2016/${GENOMEPAIR}_unfi/ "

	if [ $? -ne 0 ]
	then
		rm CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_In_Progress
		rm cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_In_Progress
		echo Failed to rsync Tumor Dat file to dback for unfi ${GENOMEPAIR} >> CNA_Manual_Fail
		echo Failed to rsync Tumor Dat file to dback for unfi ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_Fail
		echo Failed to rsync Tumor Dat file to dback for unfi ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_Fail
		echo Failed to rsync Tumor Dat file to dback for unfi ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_Fail
		echo Failed to rsync Tumor Dat file to dback for unfi ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
		exit 1
	fi

	# Exome Normal HS Metrics

	ssh ${USER}@${DATAMOVER} "rsync ${STARTDIR}/${PATIENT_NAME}/stats/${NORMALSAMPLE}.proj.md.jr.bam.picHSMetrics /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/cna_manual_2016/${GENOMEPAIR}_unfi/ "

	if [ $? -ne 0 ]
	then
	        rm CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_In_Progress
		rm cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_In_Progress
		echo Failed to rsync Normal HS metrics file to dback for unfi ${GENOMEPAIR} >> CNA_Manual_Fail
		echo Failed to rsync Normal HS metrics file to dback for unfi ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_Fail
		echo Failed to rsync Normal HS metrics file to dback for unfi ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_Fail
		echo Failed to rsync Normal HS metrics file to dback for unfi ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_Fail
		echo Failed to rsync Normal HS metrics file to dback for unfi ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
		exit 1
	fi

	# Exome Tumor HS Metrics

	ssh ${USER}@${DATAMOVER} "rsync ${STARTDIR}/${PATIENT_NAME}/stats/${TUMORSAMPLE}.proj.md.jr.bam.picHSMetrics /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/cna_manual_2016/${GENOMEPAIR}_unfi/ "

	if [ $? -ne 0 ]
	then
	        rm CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_In_Progress
		rm cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_In_Progress
		echo Failed to rsync Tumor HS metrics file to dback for unfi ${GENOMEPAIR} >> CNA_Manual_Fail
		echo Failed to rsync Tumor HS metrics file to dback for unfi ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_Fail
		echo Failed to rsync Tumor HS metrics file to dback for unfi ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_Fail
		echo Failed to rsync Tumor HS metrics file to dback for unfi ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_Fail
		echo Failed to rsync Tumor HS metrics file to dback for unfi ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
		exit 1
	fi

	# Exome HC VCF

	ssh ${USER}@${DATAMOVER} "rsync ${STARTDIR}/${PATIENT_NAME}/hc/${EXOMEPAIR}/*.HC_All.snpEff.vcf /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/cna_manual_2016/${GENOMEPAIR}_unfi/ "

	if [ $? -ne 0 ]
	then
	        rm CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_In_Progress
		rm cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_In_Progress
		echo Failed to rsync HC VCF file to dback for unfi ${GENOMEPAIR} >> CNA_Manual_Fail
		echo Failed to rsync HC VCF file to dback for unfi ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_Fail
		echo Failed to rsync HC VCF file to dback for unfi ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_Fail
		echo Failed to rsync HC VCF file to dback for unfi ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_Fail
		echo Failed to rsync HC VCF file to dback for unfi ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
		exit 1
	fi

	#}}}

	### Rsync needed files over to dback for Filtered 2012 analysis

	#{{{

	# Genome Normal Dat
	ssh ${USER}@${DATAMOVER} "rsync ${STARTDIR}/${PATIENT_NAME}/${NASSAY}/${NORMALSAMPLEG}/${NORMALDAT} /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/cna_manual_2016/${GENOMEPAIR}_filt2012/ "

	if [ $? -ne 0 ]
	then
	        rm CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_In_Progress
		rm cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_In_Progress
	        echo Failed to rsync Normal Dat file to dback for filt ${GENOMEPAIR} >> CNA_Manual_Fail
	        echo Failed to rsync Normal Dat file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_Fail
		echo Failed to rsync Normal Dat file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_Fail
	        echo Failed to rsync Normal Dat file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_Fail
	        echo Failed to rsync Normal Dat file to dback for filt ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
	        exit 1
	fi

	# Genome Tumor Dat

	ssh ${USER}@${DATAMOVER} "rsync ${STARTDIR}/${PATIENT_NAME}/${TASSAY}/${TUMORSAMPLEG}/${TUMORDAT} /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/cna_manual_2016/${GENOMEPAIR}_filt2012/ "

	if [ $? -ne 0 ]
	then
	        rm CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_In_Progress
		rm cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_In_Progress
	        echo Failed to rsync Tumor Dat file to dback for filt ${GENOMEPAIR} >> CNA_Manual_Fail
	        echo Failed to rsync Tumor Dat file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_Fail
		echo Failed to rsync Tumor Dat file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_Fail
	        echo Failed to rsync Tumor Dat file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_Fail
	        echo Failed to rsync Tumor Dat file to dback for filt ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
	        exit 1
	fi

	# Exome Normal HS Metrics

	ssh ${USER}@${DATAMOVER} "rsync ${STARTDIR}/${PATIENT_NAME}/stats/${NORMALSAMPLE}.proj.md.jr.bam.picHSMetrics /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/cna_manual_2016/${GENOMEPAIR}_filt2012/ "

	if [ $? -ne 0 ]
	then
	        rm CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_In_Progress
		rm cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_In_Progress
	        echo Failed to rsync Normal HS metrics file to dback for filt ${GENOMEPAIR} >> CNA_Manual_Fail
	        echo Failed to rsync Normal HS metrics file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_Fail
		echo Failed to rsync Normal HS metrics file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_Fail
	        echo Failed to rsync Normal HS metrics file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_Fail
	        echo Failed to rsync Normal HS metrics file to dback for filt ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
	        exit 1
	fi

	# Exome Tumor HS Metrics

	ssh ${USER}@${DATAMOVER} "rsync ${STARTDIR}/${PATIENT_NAME}/stats/${TUMORSAMPLE}.proj.md.jr.bam.picHSMetrics /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/cna_manual_2016/${GENOMEPAIR}_filt2012/ "

	if [ $? -ne 0 ]
	then
	        rm CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_In_Progress
		rm cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_In_Progress
	        echo Failed to rsync Tumor HS metrics file to dback for filt ${GENOMEPAIR} >> CNA_Manual_Fail
	        echo Failed to rsync Tumor HS metrics file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_Fail
		echo Failed to rsync Tumor HS metrics file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_Fail
	        echo Failed to rsync Tumor HS metrics file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_Fail
	        echo Failed to rsync Tumor HS metrics file to dback for filt ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
	        exit 1
	fi

	# Exome HC VCF

	ssh ${USER}@${DATAMOVER} "rsync ${STARTDIR}/${PATIENT_NAME}/hc/${EXOMEPAIR}/*.HC_All.snpEff.vcf /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/cna_manual_2016/${GENOMEPAIR}_filt2012/ "

	if [ $? -ne 0 ]
	then
	        rm CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_In_Progress
		rm cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_In_Progress
	        echo Failed to rsync HC VCF file to dback for filt ${GENOMEPAIR} >> CNA_Manual_Fail
	        echo Failed to rsync HC VCF file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_Fail
		echo Failed to rsync HS VCF file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_Fail
	        echo Failed to rsync HC VCF file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_Fail
	        echo Failed to rsync HC VCF file to dback for filt ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
	        exit 1
	fi

	#}}} 

	### Rsync needed files over to dback for Filtered 2016 analysis

	#{{{

	# Genome Normal Dat
	ssh ${USER}@${DATAMOVER} "rsync ${STARTDIR}/${PATIENT_NAME}/${NASSAY}/${NORMALSAMPLEG}/${NORMALDAT} /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/cna_manual_2016/${GENOMEPAIR}_filt2016/ "

	if [ $? -ne 0 ]
	then
	        rm CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_In_Progress
		rm cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_In_Progress
	        echo Failed to rsync Normal Dat file to dback for filt ${GENOMEPAIR} >> CNA_Manual_Fail
	        echo Failed to rsync Normal Dat file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_Fail
		echo Failed to rsync Normal Dat file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_Fail
	        echo Failed to rsync Normal Dat file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_Fail
	        echo Failed to rsync Normal Dat file to dback for filt ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
	        exit 1
	fi

	# Genome Tumor Dat

	ssh ${USER}@${DATAMOVER} "rsync ${STARTDIR}/${PATIENT_NAME}/${TASSAY}/${TUMORSAMPLEG}/${TUMORDAT} /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/cna_manual_2016/${GENOMEPAIR}_filt2016/ "

	if [ $? -ne 0 ]
	then
	        rm CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_In_Progress
		rm cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_In_Progress
	        echo Failed to rsync Tumor Dat file to dback for filt ${GENOMEPAIR} >> CNA_Manual_Fail
	        echo Failed to rsync Tumor Dat file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_Fail
		echo Failed to rsync Tumor Dat file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_Fail
	        echo Failed to rsync Tumor Dat file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_Fail
	        echo Failed to rsync Tumor Dat file to dback for filt ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
	        exit 1
	fi

	# Exome Normal HS Metrics

	ssh ${USER}@${DATAMOVER} "rsync ${STARTDIR}/${PATIENT_NAME}/stats/${NORMALSAMPLE}.proj.md.jr.bam.picHSMetrics /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/cna_manual_2016/${GENOMEPAIR}_filt2016/ "

	if [ $? -ne 0 ]
	then
	        rm CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_In_Progress
		rm cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_In_Progress
	        echo Failed to rsync Normal HS metrics file to dback for filt ${GENOMEPAIR} >> CNA_Manual_Fail
	        echo Failed to rsync Normal HS metrics file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_Fail
		echo Failed to rsync Normal HS metrics file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_Fail
	        echo Failed to rsync Normal HS metrics file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_Fail
	        echo Failed to rsync Normal HS metrics file to dback for filt ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
	        exit 1
	fi

	# Exome Tumor HS Metrics

	ssh ${USER}@${DATAMOVER} "rsync ${STARTDIR}/${PATIENT_NAME}/stats/${TUMORSAMPLE}.proj.md.jr.bam.picHSMetrics /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/cna_manual_2016/${GENOMEPAIR}_filt2016/ "

	if [ $? -ne 0 ]
	then
	        rm CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_In_Progress
		rm cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_In_Progress
	        echo Failed to rsync Tumor HS metrics file to dback for filt ${GENOMEPAIR} >> CNA_Manual_Fail
	        echo Failed to rsync Tumor HS metrics file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_Fail
		echo Failed to rsync Tumor HS metrics file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_Fail
	        echo Failed to rsync Tumor HS metrics file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_Fail
	        echo Failed to rsync Tumor HS metrics file to dback for filt ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
	        exit 1
	fi

	# Exome HC VCF

	ssh ${USER}@${DATAMOVER} "rsync ${STARTDIR}/${PATIENT_NAME}/hc/${EXOMEPAIR}/*.HC_All.snpEff.vcf /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/cna_manual_2016/${GENOMEPAIR}_filt2016/ "

	if [ $? -ne 0 ]
	then
	        rm CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_In_Progress
		rm cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_In_Progress
	        echo Failed to rsync HC VCF file to dback for filt ${GENOMEPAIR} >> CNA_Manual_Fail
	        echo Failed to rsync HC VCF file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_Fail
		echo Failed to rsync HC VCF file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_Fail
	        echo Failed to rsync HC VCF file to dback for filt ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_Fail
	        echo Failed to rsync HC VCF file to dback for filt ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
	        exit 1
	fi
	
	#}}}

	### Start pbs jobs on PNAP

	#{{{ 

	DIR=/scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/cna_manual_2016/${GENOMEPAIR}_unfi

	# Unfi CNA
	ssh ${USER}@dback-login1.tgen.org "source /etc/profile ; cd ${DIR} ; sbatch -n 1 -N 1 --mem-per-cpu 3000 --cpus-per-task 4 --output ${HOME}/STD_OUT/%x_${GENOMEPAIR}_%j.out --export DATAMOVERIP=${DATAMOVERIP},DATAMOVER=${DATAMOVER},BASEDIR=${BASEDIR},JOBSDIR=${JOBSDIR},STARTDIR="${STARTDIR}",BEDFILE=${BEDFILE},GENOMEPAIR=${GENOMEPAIR},PATIENT_NAME="${PATIENT_NAME}",DIR=${DIR},NORMALSAMPLE=${NORMALSAMPLE},TUMORSAMPLE=${TUMORSAMPLE},VCF="${EXOMEPAIR}.HC_All.snpEff.vcf",NORMALDAT="${NORMALSAMPLEG}.proj.md.jr.bam.clc.cln.dat",TUMORDAT="${TUMORSAMPLEG}.proj.md.jr.bam.clc.cln.dat",NORMALSAMPLEG="${NORMALSAMPLEG}",TUMORSAMPLEG="${TUMORSAMPLEG}",OFILE=${GENOMEPAIR},EXPECTEDPAIRS=${EXPECTEDPAIRS} ${JOBSDIR}/ngs_cna2015_WGcenterWithExomes_V4.pbs "

	if [ $? -ne 0 ]
	then
		rm CNA_Manual_In_Progress
        	rm cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_In_Progress
		rm cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_In_Progress
        	rm cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_In_Progress
		echo Failed to Start unfi qsub job for ${GENOMEPAIR} >> CNA_Manual_Fail
		echo Failed to Start unfi qsub job for ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_Fail
		echo Failed to Start unfi qsub job for ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_Fail
		echo Failed to Start unfi qsub job for ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_unfi/CNA_Manual_Fail
		echo Failed to Start unfi qsub job for ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org 	
		exit 1
	fi

	DIR=/scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/cna_manual_2016/${GENOMEPAIR}_filt2012

	# Filt CNA
	ssh ${USER}@dback-login1.tgen.org "source /etc/profile ; cd ${DIR} ; sbatch -n 1 -N 1 --mem-per-cpu 3000 --cpus-per-task 4 --output ${HOME}/STD_OUT/%x_${GENOMEPAIR}_%j.out --export DATAMOVERIP=${DATAMOVERIP},DATAMOVER=${DATAMOVER},BASEDIR=${BASEDIR},JOBSDIR=${JOBSDIR},STARTDIR="${STARTDIR}",BEDFILE=${BEDFILE},GENOMEPAIR=${GENOMEPAIR},PATIENT_NAME="${PATIENT_NAME}",DIR=${DIR},NORMALSAMPLE=${NORMALSAMPLE},TUMORSAMPLE=${TUMORSAMPLE},VCF="${EXOMEPAIR}.HC_All.snpEff.vcf",NORMALDAT="${NORMALSAMPLEG}.proj.md.jr.bam.clc.cln.dat",TUMORDAT="${TUMORSAMPLEG}.proj.md.jr.bam.clc.cln.dat",NORMALSAMPLEG="${NORMALSAMPLEG}",TUMORSAMPLEG="${TUMORSAMPLEG}",OFILE=${GENOMEPAIR},EXPECTEDPAIRS=${EXPECTEDPAIRS} ${JOBSDIR}/ngs_cna2015_WGcenterWithExomesFilt2012_V3.pbs "

	if [ $? -ne 0 ]
	then
	        rm CNA_Manual_In_Progress
		rm cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_In_Progress
		rm cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_In_Progress
		echo Failed to Start filt qsub job for ${GENOMEPAIR} >> CNA_Manual_Fail
		echo Failed to Start filt qsub job for ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2012/CNA_Manual_Fail
		echo Failed to Start filt qsub job for ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_Fail
		echo Failed to Start filt qsub job for ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
	fi

	DIR=/scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/cna_manual_2016/${GENOMEPAIR}_filt2016

	# Filt CNA
	ssh ${USER}@dback-login1.tgen.org "source /etc/profile ; cd ${DIR} ; sbatch -n 1 -N 1 --mem-per-cpu 3000 --cpus-per-task 4 --output ${HOME}/STD_OUT/%x_${GENOMEPAIR}_%j.out --export DATAMOVERIP=${DATAMOVERIP},DATAMOVER=${DATAMOVER},BASEDIR=${BASEDIR},JOBSDIR=${JOBSDIR},STARTDIR="${STARTDIR}",BEDFILE=${BEDFILE},GENOMEPAIR=${GENOMEPAIR},PATIENT_NAME="${PATIENT_NAME}",DIR=${DIR},NORMALSAMPLE=${NORMALSAMPLE},TUMORSAMPLE=${TUMORSAMPLE},VCF="${EXOMEPAIR}.HC_All.snpEff.vcf",NORMALDAT="${NORMALSAMPLEG}.proj.md.jr.bam.clc.cln.dat",TUMORDAT="${TUMORSAMPLEG}.proj.md.jr.bam.clc.cln.dat",NORMALSAMPLEG="${NORMALSAMPLEG}",TUMORSAMPLEG="${TUMORSAMPLEG}",OFILE=${GENOMEPAIR},EXPECTEDPAIRS=${EXPECTEDPAIRS} ${JOBSDIR}/ngs_cna2015_WGcenterWithExomesFilt2016_V3.pbs "

	if [ $? -ne 0 ]
	then
	        rm CNA_Manual_In_Progress
	        rm cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_In_Progress
	        echo Failed to Start filt qsub job for ${GENOMEPAIR} >> CNA_Manual_Fail
	        echo Failed to Start filt qsub job for ${GENOMEPAIR} >> cna_manual_2016/${GENOMEPAIR}_filt2016/CNA_Manual_Fail
	        echo Failed to Start filt qsub job for ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
	fi

	#}}}

elif [[ ${DELLY} == "Yes" ]]
then
	# Make needed directories for delly

	DIR=/scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/delly/${GENOMEPAIR}

	ssh ${USER}@${DATAMOVER} "mkdir -p /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/delly/${GENOMEPAIR}"

	if [ $? -ne 0 ]
	then
	        rm Delly_In_Progress
		rm delly/${GENOMEPAIR}/Delly_In_Progress
		echo Failed to make delly dir on dback for ${GENOMEPAIR} >> Delly_Fail
		echo Failed to make delly dir on dback for ${GENOMEPAIR} >> delly/${GENOMEPAIR}/Delly_Fail
		echo Failed to make delly dir on dback for ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
	        exit 1
	fi

	### Rsync needed files over to dback for delly analysis

	ssh ${USER}@${DATAMOVER} " if [ -d \"\` ls -d /scratch/tgenjetstream/projects/${PATIENT_NAME}* | tail -n1 \`\" ] ; then LNDIR=\`ls -d /scratch/tgenjetstream/projects/${PATIENT_NAME}* | tail -n1 \` ; ln -s \${LNDIR}/${TASSAY}/${TUMORSAMPLEG}/${TUMORSAMPLEG}.bwa.final.bam /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/delly/${GENOMEPAIR}/${TUMORSAMPLEG}.bwa.final.bam ; else rsync ${STARTDIR}/${PATIENT_NAME}/${TASSAY}/${TUMORSAMPLEG}/${TUMORSAMPLEG}.bwa.final.bam /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/delly/${GENOMEPAIR}/  ; fi "

	if [ $? -ne 0 ]
	then
	        rm Delly_In_Progress
	        rm delly/${GENOMEPAIR}/Delly_In_Progress
	        echo Failed to rsync delly bam for ${GENOMEPAIR} >> Delly_Fail
	        echo Failed to rsync delly bam for ${GENOMEPAIR} >> delly/${GENOMEPAIR}/Delly_Fail
	        echo Failed to rsync delly bam for ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
        	exit 1
	fi

	ssh ${USER}@${DATAMOVER} " if [ -d \"\` ls -d /scratch/tgenjetstream/projects/${PATIENT_NAME}* | tail -n1 \`\" ] ; then LNDIR=\`ls -d /scratch/tgenjetstream/projects/${PATIENT_NAME}* | tail -n1 \` ; ln -s \${LNDIR}/${TASSAY}/${TUMORSAMPLEG}/${TUMORSAMPLEG}.bwa.final.bai /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/delly/${GENOMEPAIR}/${TUMORSAMPLEG}.bwa.final.bai ; else rsync ${STARTDIR}/${PATIENT_NAME}/${TASSAY}/${TUMORSAMPLEG}/${TUMORSAMPLEG}.bwa.final.bai /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/delly/${GENOMEPAIR}/  ; fi   "

	if [ $? -ne 0 ]
	then
	        rm Delly_In_Progress
	        rm delly/${GENOMEPAIR}/Delly_In_Progress
	        echo Failed to rsync delly bam for ${GENOMEPAIR} >> Delly_Fail
	        echo Failed to rsync delly bam for ${GENOMEPAIR} >> delly/${GENOMEPAIR}/Delly_Fail
	        echo Failed to rsync delly bam for ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
	        exit 1
	fi



	ssh ${USER}@${DATAMOVER} " if [ -d \"\` ls -d /scratch/tgenjetstream/projects/${PATIENT_NAME}* | tail -n1 \`\" ] ; then LNDIR=\`ls -d /scratch/tgenjetstream/projects/${PATIENT_NAME}* | tail -n1\` ; ln -s \${LNDIR}/${NASSAY}/${NORMALSAMPLEG}/${NORMALSAMPLEG}.bwa.final.bam /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/delly/${GENOMEPAIR}/${NORMALSAMPLEG}.bwa.final.bam ; else rsync ${STARTDIR}/${PATIENT_NAME}/${NASSAY}/${NORMALSAMPLEG}/${NORMALSAMPLEG}.bwa.final.bam /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/delly/${GENOMEPAIR}/  ; fi "

	if [ $? -ne 0 ]
	then
	        rm Delly_In_Progress
	        rm delly/${GENOMEPAIR}/Delly_In_Progress
	        echo Failed to rsync delly bam for ${GENOMEPAIR} >> Delly_Fail
	        echo Failed to rsync delly bam for ${GENOMEPAIR} >> delly/${GENOMEPAIR}/Delly_Fail
	        echo Failed to rsync delly bam for ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
	        exit 1
	fi

	ssh ${USER}@${DATAMOVER} " if [ -d \"\` ls -d /scratch/tgenjetstream/projects/${PATIENT_NAME}* | tail -n1 \`\" ] ; then LNDIR=\`ls -d /scratch/tgenjetstream/projects/${PATIENT_NAME}* | tail -n1\` ; ln -s \${LNDIR}/${NASSAY}/${NORMALSAMPLEG}/${NORMALSAMPLEG}.bwa.final.bai /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/delly/${GENOMEPAIR}/${NORMALSAMPLEG}.bwa.final.bai ; else rsync ${STARTDIR}/${PATIENT_NAME}/${NASSAY}/${NORMALSAMPLEG}/${NORMALSAMPLEG}.bwa.final.bai /scratch/${USER}/post_medusa_processing/${PATIENT_NAME}/delly/${GENOMEPAIR}/ ; fi   "

	if [ $? -ne 0 ]
	then
	        rm Delly_In_Progress
        	rm delly/${GENOMEPAIR}/Delly_In_Progress
        	echo Failed to rsync delly bam for ${GENOMEPAIR} >> Delly_Fail
        	echo Failed to rsync delly bam for ${GENOMEPAIR} >> delly/${GENOMEPAIR}/Delly_Fail
        	echo Failed to rsync delly bam for ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
        	exit 1
	fi

	# Start Delly PBS

	ssh ${USER}@dback-login1.tgen.org "source /etc/profile ; cd ${DIR} ; sbatch -n 1 -N 1 --mem-per-cpu 3000 --cpus-per-task 8 --output ${HOME}/STD_OUT/%x_${GENOMEPAIR}_%j.out --export DATAMOVERIP=${DATAMOVERIP},DATAMOVER=${DATAMOVER},BASEDIR=${BASEDIR},JOBSDIR=${JOBSDIR},DIR=${DIR},STARTDIR="${STARTDIR}",GENOMEPAIR="${GENOMEPAIR}",PATIENT_NAME="${PATIENT_NAME}",TUMORSAMPLEG="${TUMORSAMPLEG}",NORMALSAMPLEG="${NORMALSAMPLEG}",EXPECTEDPAIRS=${EXPECTEDPAIRS} ${JOBSDIR}/delly.pbs " 

	if [ $? -ne 0 ]
	then
		rm Delly_In_Progress
		rm delly/${GENOMEPAIR}/Delly_In_Progress
		echo Failed to start sbatch job for ${GENOMEPAIR} >> Delly_Fail
		echo Failed to start sbatch job for ${GENOMEPAIR} >> delly/${GENOMEPAIR}/Delly_Fail
		echo Failed to start sbatch job for ${GENOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org
		exit 1
	fi
fi




