#!/bin/bash
#PBS -l nodes=1:ppn=16
#PBS -N Salmon_GTF
#PBS -l walltime=48:00:00
#PBS -j oe
#PBS -o ${HOME}/STD_OUT/${PBS_JOBNAME}_${RNASAMPLEID}_${PBS_JOBID}.out

cd $PBS_O_WORKDIR

# Load old salmon to get path to libraries. Needed until IT gets the new salmon installed as a package.
module load salmon/0.5.0

#####################################################
##
##   VARIABLES
##
#####################################################

SALMON_BINARY=/home/clegendre/bin/salmon/0.5.1/bin/salmon
SALMON_INDEX_GTF=/home/achristofferson/salmon_index/Homo_sapiens.GRCh37.74.gtf.hs37d5.EGFRvIII.fa.index
GTF=/home/tgenref/pecan/ensembl_v74/Homo_sapiens.GRCh37.74.gtf.hs37d5.EGFRvIII.gtf

# Generate a variable with each R1 and R2 fastq in the same order

#Read1
#Add loop control for first line processed
CONTROL=1
for line in `ls ${SAMPLE}*_R1_001.fastq.gz`
do
        if [ ${CONTROL} -eq 1 ]
        then
                echo First read is ${line}
                FORWARD="${line}"
                CONTROL=0
        else
                echo Additional read is ${line}
                FORWARD="${FORWARD} ${line}"
        fi
done

echo Input list of forward reads:
echo ${FORWARD}

#Read2
#Reset loop control for first line processed
CONTROL=1
for line in `ls ${SAMPLE}*_R2_001.fastq.gz`
do
        if [ ${CONTROL} -eq 1 ]
        then
                echo First read is ${line}
                REVERSE="${line}"
                CONTROL=0
        else
                echo Additional read is ${line}
                REVERSE="${REVERSE} ${line}"
        fi
done

echo Input list of reverse reads:
echo ${REVERSE}


${SALMON_BINARY} quant \
        --index ${SALMON_INDEX_GTF} \
        --libType ${RNATYPE} \
        --numBootstraps 100 \
        --mates1 <(zcat ${FORWARD}) \
        --mates2 <(zcat ${REVERSE}) \
        --threads 14 \
        --geneMap ${GTF} \
        --output ${SAMPLE}.salmonDir/ensembl74_GTF

if [ "$?" -ne 0 ]
then
        ssh ${USER}@172.19.1.222 "cd ${STARTDIR}/${PATIENT_NAME} ; \
                rm ${ASSAY}/${RNASAMPLEID}/${RNASAMPLEID}.salmonDir/ensembl74_GTF/Salmon_In_Progress ; \
                echo Salmon failed for ${RNASAMPLEID} >> ${ASSAY}/${RNASAMPLEID}/${RNASAMPLEID}.salmonDir/ensembl74_GTF/Salmon_Fail ; \
                echo Salmon failed for ${RNASAMPLEID} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org"
	exit 1
else
	#Rename Result files
	cd ${SAMPLE}.salmonDir/ensembl74_GTF
	mv quant.genes.sf ${SAMPLE}_salmon_e74GTF_genes.sf
	mv quant.sf ${SAMPLE}_salmon_e74GTF_transcripts.sf
	mv quant_bootstraps.sf ${SAMPLE}_salmon_e74GTF_quant_bootstraps.sf
	mv stats.tsv ${SAMPLE}_salmon_e74GTF_stats.tsv

	DIR=`pwd`

	ssh ${USER}@172.19.1.222 "rsync -r ${DIR} ${STARTDIR}/${PATIENT_NAME}/${ASSAY}/${RNASAMPLEID}/${RNASAMPLEID}.salmonDir/ "

        if [ $? -ne 0 ]
        then
                ssh ${USER}@172.19.1.222 "rm ${STARTDIR}/${PATIENT_NAME}/${ASSAY}/${RNASAMPLEID}/${RNASAMPLEID}.salmonDir/ensembl74_GTF/Salmon_In_Progress ; \
                        echo Failed to rsync salmon dir to isilon for ${RNASAMPLEID} >> ${STARTDIR}/${PATIENT_NAME}/${ASSAY}/${RNASAMPLEID}/${RNASAMPLEID}.salmonDir/ensembl74_GTF/Salmon_Fail ; \
                        echo Failed to rsync salmon dir to isilon for ${RNASAMPLEID} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org"

        else
                ssh ${USER}@172.19.1.222 "rm ${STARTDIR}/${PATIENT_NAME}/${ASSAY}/${RNASAMPLEID}/${RNASAMPLEID}.salmonDir/ensembl74_GTF/Salmon_In_Progress ; \
                        touch ${STARTDIR}/${PATIENT_NAME}/${ASSAY}/${RNASAMPLEID}/${RNASAMPLEID}.salmonDir/ensembl74_GTF/Salmon_Complete ; \
                        COMPLETE=0 ; \
                        for RSAMPLE in \`echo ${RNACHECK} |sed 's/@/\n/g'\`
                        do
				echo ${STARTDIR}/${PATIENT_NAME}/\${RSAMPLE}.salmonDir/ensembl74_GTF/Salmon_Complete
                                if [ ! -f ${STARTDIR}/${PATIENT_NAME}/\${RSAMPLE}.salmonDir/ensembl74_cDNA/Salmon_Complete ] || [ ! -f ${STARTDIR}/${PATIENT_NAME}/\${RSAMPLE}.salmonDir/ensembl74_GTF/Salmon_Complete ] || [ ! -f ${STARTDIR}/${PATIENT_NAME}/\${RSAMPLE}.salmonDir/ensembl74_GTF_V7.2/Salmon_Complete ]
                                then
                                        COMPLETE=1
                                fi
                        done ; \
                        if [ \${COMPLETE} = 0 ]
                        then
                                rm ${STARTDIR}/${PATIENT_NAME}/Salmon_In_Progress
                                touch ${STARTDIR}/${PATIENT_NAME}/Salmon_Complete
                        fi "



        fi
fi



