#!/bin/bash
#PBS -l nodes=1:ppn=16
#PBS -N snpEFF
#PBS -l walltime=24:00:00
#PBS -j oe
#PBS -o ${HOME}/STD_OUT/${PBS_JOBNAME}_${TUMOR}_${PBS_JOBID}.out

cd $PBS_O_WORKDIR
cd ${DIR}

# Tool and Data Base Paths
SNPEFFPATH=/home/achristofferson/local/snpEff_4.2_2015-12-05
SNPSIFT=/home/achristofferson/local/snpEff_4.2_2015-12-05/SnpSift.jar
SNPEFF=/home/achristofferson/local/snpEff_4.2_2015-12-05/snpEff.jar
VCFVALIDATOR=/home/achristofferson/local/vcftools_v0.1.14/bin/vcf-validator
BCFTOOLS=/home/achristofferson/local/bcftools_v1.2-172-g06ce5a1/bcftools
HEADER=/home/tgenref/pecan/post_central_pipe_processing/post_medusa_V1.0/vcf_RNA_INFO_header.hdr

# dbNSFP DB (will need to change when transfered to root)
DBNSFP=/home/tgenref/pecan/post_central_pipe_processing/post_medusa_V1.0/dbNSFP2.9/dbNSFP2.9.txt.gz
DBNSFP=/home/tgenref/pecan/dbNSFP2.9/dbNSFP2.9.txt.gz

# vcf file names
ALL="`ls ${PAIR}.merged.allTranscripts*final.vcf`"
CANON="`ls ${PAIR}.merged.canonicalOnly*final.vcf`"

if [ "`grep -f ${HEADER} ${ALL} | wc -l`" = 3  ]
then
	ARNA="Yes"
else
	ARNA="No"
fi

if [ "`grep -f ${HEADER} ${CANON} | wc -l`" = 3  ]
then
	CRNA="Yes"
else
	CRNA="No"
fi

#############################
#
# Update All Transcripts VCF
#
#############################

# Remove EFF and dbNSFP annotations 

java -jar ${SNPSIFT} rmInfo ${ALL} INDEL LOF NMD EFF dbNSFP_GERP++_RS dbNSFP_LRT_score dbNSFP_MutationAssessor_score dbNSFP_FATHMM_score dbNSFP_GERP++_NR dbNSFP_Polyphen2_HDIV_score dbNSFP_Polyphen2_HVAR_pred dbNSFP_SIFT_score dbNSFP_Polyphen2_HVAR_score dbNSFP_MutationTaster_score dbNSFP_Interpro_domain | \
	grep -v "##INFO=<ID=dbNSFP_" | \
	grep -v "##SnpSiftVersion" | \
	grep -v "##SnpEffVersion" | \
	grep -v "##SnpEffCmd" | \
	grep -v "##INFO=<ID=INDEL" | \
	grep -v "##INFO=<ID=EFF" | \
	grep -v "##INFO=<ID=LOF" | \
	grep -v "##INFO=<ID=NMD" | \
	grep -v "##INFO=<ID=SEURAT_IC," | \
	grep -v "##INFO=<ID=SEURAT_IHP," | \
	grep -v "##INFO=<ID=SEURAT_NT," | \
	grep -v "##INFO=<ID=SEURAT_OVERLAP," | \
	grep -v "##INFO=<ID=SEURAT_QSI," | \
	grep -v "##INFO=<ID=SEURAT_QSI_NT," | \
	grep -v "##INFO=<ID=SEURAT_RC," | \
	grep -v "##INFO=<ID=SEURAT_RU," | \
	grep -v "##INFO=<ID=SEURAT_SGT," | \
	grep -v "##INFO=<ID=SEURAT_SOMATIC," | \
	grep -v "##INFO=<ID=SEURAT_TQSI," | \
	grep -v "##INFO=<ID=SEURAT_TQSI_NT," | \
	grep -v "##FILTER=<ID=REJECT" | \
	sed 's|##FORMAT=<ID=MUTECT_AD,Number=.,Type=Integer,Description="Not provided in original VCF header">|##FORMAT=<ID=MUTECT_AD,Number=.,Type=Integer,Description="Allelic depths for the ref and alt alleles in the order listed">|g' | \
	sed 's|##FORMAT=<ID=MUTECT_BQ,Number=A,Type=Float,Description="Not provided in original VCF header">|##FORMAT=<ID=MUTECT_BQ,Number=A,Type=Float,Description="Average base quality for reads supporting alleles">|g' | \
	sed 's|##FORMAT=<ID=MUTECT_DP,Number=1,Type=Integer,Description="Not provided in original VCF header">|##FORMAT=<ID=MUTECT_DP,Number=1,Type=Integer,Description="Approximate read depth (reads with MQ=255 or with bad mates are filtered)">|g' | \
	sed 's|##FORMAT=<ID=MUTECT_FA,Number=A,Type=Float,Description="Not provided in original VCF header">|##FORMAT=<ID=MUTECT_FA,Number=A,Type=Float,Description="Allele fraction of the alternate allele with regard to reference">|g' | \
	sed 's|##FORMAT=<ID=MUTECT_GQ,Number=1,Type=Integer,Description="Not provided in original VCF header">|##FORMAT=<ID=MUTECT_GQ,Number=1,Type=Integer,Description="Genotype Quality">|g' | \
	sed 's|##FORMAT=<ID=MUTECT_GT,Number=1,Type=String,Description="Not provided in original VCF header">|##FORMAT=<ID=MUTECT_GT,Number=1,Type=String,Description="Genotype">|g' | \
	sed 's|##FORMAT=<ID=MUTECT_PL,Number=G,Type=Integer,Description="Not provided in original VCF header">|##FORMAT=<ID=MUTECT_PL,Number=G,Type=Integer,Description="Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification">|g' | \
	sed 's|##FORMAT=<ID=MUTECT_SS,Number=1,Type=Integer,Description="Not provided in original VCF header">|##FORMAT=<ID=MUTECT_SS,Number=1,Type=Integer,Description="Variant status relative to non-adjacent Normal,0=wildtype,1=germline,2=somatic,3=LOH,4=post-transcriptional modification,5=unknown">|g' | \
	sed 's|##INFO=<ID=MUTECT_DB,Number=0,Type=Flag,Description="Not provided in original VCF header">|##INFO=<ID=MUTECT_DB,Number=0,Type=Flag,Description="dbSNP Membership">|g' | \
	sed 's|##INFO=<ID=MUTECT_MQ0,Number=1,Type=Integer,Description="Not provided in original VCF header">|##INFO=<ID=MUTECT_MQ0,Number=1,Type=Integer,Description="Total Mapping Quality Zero Reads">|g' | \
	sed 's|##INFO=<ID=MUTECT_SOMATIC,Number=0,Type=Flag,Description="Not provided in original VCF header">|##INFO=<ID=MUTECT_SOMATIC,Number=0,Type=Flag,Description="Somatic event">|g' | \
	sed 's|##INFO=<ID=MUTECT_VT,Number=1,Type=String,Description="Not provided in original VCF header">|##INFO=<ID=MUTECT_VT,Number=1,Type=String,Description="Variant type, can be SNP, INS or DEL">|g' > ${ALL}.noEFF

if [ $? -ne 0 ]
then
        rm snpEFF_on_merged_allTranscripts_in_progress
        touch snpEFF_on_merged_allTranscripts_Failed
        rm ${ALL}.noEFF
        echo "### ${ALL} failed at Step 1) rmInfo" >> snpEFF_on_merged_allTranscripts_Failed
fi

# Add SnpEff ANN annotations (default with version 4.2)

java -Xmx14g -jar ${SNPEFF} -c ${SNPEFFPATH}/snpEff.config -lof GRCh37.74 ${ALL}.noEFF | \
	java -jar ${SNPSIFT} varType - | \
	grep -v "##INFO=<ID=HOM,Number=A,Type=Flag,Description=\"Variant is homozygous\">" | \
	grep -v "##INFO=<ID=HET,Number=A,Type=Flag,Description=\"Variant is heterozygous\">" > ${ALL}.ANN

if [ $? -ne 0 ]
then
        rm snpEFF_on_merged_allTranscripts_in_progress
        touch snpEFF_on_merged_allTranscripts_Failed
        rm ${ALL}.ANN
	rm ${ALL}.noEFF
        echo "### ${ALL} failed at Step 2) SnpEff" >> snpEFF_on_merged_allTranscripts_Failed
fi

# Add dbNSFP annotations

java -jar ${SNPSIFT} dbnsfp \
	-v \
	-a \
	-db ${DBNSFP} \
	-f CADD_raw,CADD_raw_rankscore,CADD_phred,Interpro_domain,Polyphen2_HVAR_pred,GERP++_NR,GERP++_RS,LRT_score,MutationTaster_score,MutationAssessor_score,FATHMM_score,Polyphen2_HVAR_score,SIFT_score,Polyphen2_HDIV_score,MetaSVM_score,MetaSVM_rankscore,MetaSVM_pred,MetaLR_score,MetaLR_rankscore,MetaLR_pred,Reliability_index \
	${ALL}.ANN > ${ALL}.dbNSFP

if [ $? -ne 0 ]
then
        rm snpEFF_on_merged_allTranscripts_in_progress
        touch snpEFF_on_merged_allTranscripts_Failed
        echo "### ${ALL} failed at Step 3) dbNSFP" >> snpEFF_on_merged_allTranscripts_Failed
        rm snpEff_genes.txt
        rm snpEff_summary.html
        rm ${ALL}.noEFF
        rm ${ALL}.ANN
        rm ${ALL}.dbNSFP
fi

# Fix header lines that are missing, have spaces, dbNSFP errors, RNA_ALT_FREQ float problem.

if [ ${ARNA} = "No"  ]
then
	${BCFTOOLS} annotate -h ${HEADER} ${ALL}.dbNSFP | \
	sed 's/'\'' '\"'/'\'\"'/g' | \
	sed 's|##FILTER=<ID=PASS>|##FILTER=<ID=PASS,Description="Accept as a confident somatic mutation">|g' | \
	sed 's|##FILTER=<ID=PASS,Description="All filters passed">|##FILTER=<ID=PASS,Description="Accept as a confident somatic mutation">|g' | \
	sed 's/dbNSFP_CADD_raw,Number=A/dbNSFP_CADD_raw,Number=./g' | \
	sed 's/dbNSFP_CADD_raw_rankscore,Number=A/dbNSFP_CADD_raw_rankscore,Number=./g' | \
	sed 's/dbNSFP_CADD_phred,Number=A/dbNSFP_CADD_phred,Number=./g' | \
	sed 's/dbNSFP_Polyphen2_HVAR_pred,Number=A/dbNSFP_Polyphen2_HVAR_pred,Number=./g' | \
	sed 's/dbNSFP_GERP___NR,Number=A/dbNSFP_GERP___NR,Number=./g' | \
	sed 's/dbNSFP_GERP___RS,Number=A/dbNSFP_GERP___RS,Number=./g' | \
	sed 's/dbNSFP_LRT_score,Number=A/dbNSFP_LRT_score,Number=./g' | \
	sed 's/dbNSFP_MutationTaster_score,Number=A/dbNSFP_MutationTaster_score,Number=./g' | \
	sed 's/dbNSFP_MutationAssessor_score,Number=A/dbNSFP_MutationAssessor_score,Number=./g' | \
	sed 's/dbNSFP_MetaLR_rankscore,Number=A/dbNSFP_MetaLR_rankscore,Number=./g' | \
	sed 's/dbNSFP_MetaSVM_rankscore,Number=A/dbNSFP_MetaSVM_rankscore,Number=./g' | \
	sed 's/dbNSFP_MetaSVM_pred,Number=A/dbNSFP_MetaSVM_pred,Number=./g' | \
	sed 's/dbNSFP_MetaSVM_score,Number=A/dbNSFP_MetaSVM_score,Number=./g' | \
	sed 's/dbNSFP_MetaLR_pred,Number=A/dbNSFP_MetaLR_pred,Number=./g' | \
	sed 's/dbNSFP_MetaLR_score,Number=A/dbNSFP_MetaLR_score,Number=./g' | \
	sed 's/dbNSFP_Reliability_index,Number=A/dbNSFP_Reliability_index,Number=./g' | \
	sed 's/dbNSFP_Polyphen2_HVAR_score,Number=A/dbNSFP_Polyphen2_HVAR_score,Number=./g' | \
        sed 's/dbNSFP_SIFT_score,Number=A/dbNSFP_SIFT_score,Number=./g' | \
        sed 's/dbNSFP_Polyphen2_HVAR_pred,Number=A/dbNSFP_Polyphen2_HVAR_pred,Number=./g' | \
        sed 's/dbNSFP_LRT_score,Number=A/dbNSFP_LRT_score,Number=./g' | \
        sed 's/dbNSFP_FATHMM_score,Number=A/dbNSFP_FATHMM_score,Number=./g' | \
        sed 's/dbNSFP_Polyphen2_HDIV_score,Number=A/dbNSFP_Polyphen2_HDIV_score,Number=./g' | \
        sed 's/dbNSFP_Interpro_domain,Number=A/dbNSFP_Interpro_domain,Number=./g' | \
        sed 's/RNA_ALT_FREQ,Number=1,Type=Integer/RNA_ALT_FREQ,Number=1,Type=Float/g' | \
        sed 's/RNA_ALT_COUNT,Number=1,Type=Float/RNA_ALT_COUNT,Number=1,Type=Integer/g' | \
	sed 's|##reference=file:///home/tgenref/pecan/refgenome/hs37d5/hs37d5_plusRibo_plusOncoViruses_plusERCC.fa|##reference=file:///home/tgenref/pecan/bwa_index/hs37d5_plusRibo_plusOncoViruses_plusERCC.fa|g' > ${ALL}.FIX 

	if [ $? -ne 0 ]
	then
		rm snpEFF_on_merged_allTranscripts_in_progress
        	touch snpEFF_on_merged_allTranscripts_Failed
        	echo "### ${ALL} failed at Step 4) FIX" >> snpEFF_on_merged_allTranscripts_Failed
        	rm snpEff_genes.txt
        	rm snpEff_summary.html
        	rm ${ALL}.noEFF
        	rm ${ALL}.ANN
        	rm ${ALL}.dbNSFP
        	rm ${ALL}.FIX
	else
		VALIDATOR="`${VCFVALIDATOR} ${ALL}.FIX | wc -l`"
                if [ ${VALIDATOR} = "0" ]
                then
                        rm snpEFF_on_merged_allTranscripts_in_progress
                        touch snpEFF_on_merged_allTranscripts_Complete
                        rm snpEff_genes.txt
                        rm snpEff_summary.html
                        rm ${ALL}.noEFF
                        rm ${ALL}.ANN
                        rm ${ALL}.dbNSFP
                        mv ${ALL}.FIX ${ALL}
                else
                        rm snpEFF_on_merged_allTranscripts_in_progress
                        touch snpEFF_on_merged_allTranscripts_Failed
                        echo "### ${ALL} failed at Step 5) Validation" >> snpEFF_on_merged_allTranscripts_Failed
                        rm snpEff_genes.txt
                        rm snpEff_summary.html
                        rm ${ALL}.noEFF
                        rm ${ALL}.ANN
                        rm ${ALL}.dbNSFP
                        rm ${ALL}.FIX
                fi	
	fi
else
	sed 's/'\'' '\"'/'\'\"'/g' ${ALL}.dbNSFP | \
	sed 's|##FILTER=<ID=PASS>|##FILTER=<ID=PASS,Description="Accept as a confident somatic mutation">|g' | \
	sed 's/dbNSFP_CADD_raw,Number=A/dbNSFP_CADD_raw,Number=./g' | \
        sed 's/dbNSFP_CADD_raw_rankscore,Number=A/dbNSFP_CADD_raw_rankscore,Number=./g' | \
        sed 's/dbNSFP_CADD_phred,Number=A/dbNSFP_CADD_phred,Number=./g' | \
        sed 's/dbNSFP_Polyphen2_HVAR_pred,Number=A/dbNSFP_Polyphen2_HVAR_pred,Number=./g' | \
        sed 's/dbNSFP_GERP___NR,Number=A/dbNSFP_GERP___NR,Number=./g' | \
        sed 's/dbNSFP_GERP___RS,Number=A/dbNSFP_GERP___RS,Number=./g' | \
        sed 's/dbNSFP_LRT_score,Number=A/dbNSFP_LRT_score,Number=./g' | \
        sed 's/dbNSFP_MutationTaster_score,Number=A/dbNSFP_MutationTaster_score,Number=./g' | \
        sed 's/dbNSFP_MutationAssessor_score,Number=A/dbNSFP_MutationAssessor_score,Number=./g' | \
        sed 's/dbNSFP_MetaLR_rankscore,Number=A/dbNSFP_MetaLR_rankscore,Number=./g' | \
        sed 's/dbNSFP_MetaSVM_rankscore,Number=A/dbNSFP_MetaSVM_rankscore,Number=./g' | \
        sed 's/dbNSFP_MetaSVM_pred,Number=A/dbNSFP_MetaSVM_pred,Number=./g' | \
        sed 's/dbNSFP_MetaSVM_score,Number=A/dbNSFP_MetaSVM_score,Number=./g' | \
        sed 's/dbNSFP_MetaLR_pred,Number=A/dbNSFP_MetaLR_pred,Number=./g' | \
        sed 's/dbNSFP_MetaLR_score,Number=A/dbNSFP_MetaLR_score,Number=./g' | \
        sed 's/dbNSFP_Reliability_index,Number=A/dbNSFP_Reliability_index,Number=./g' | \
        sed 's/dbNSFP_Polyphen2_HVAR_score,Number=A/dbNSFP_Polyphen2_HVAR_score,Number=./g' | \
        sed 's/dbNSFP_SIFT_score,Number=A/dbNSFP_SIFT_score,Number=./g' | \
        sed 's/dbNSFP_Polyphen2_HVAR_pred,Number=A/dbNSFP_Polyphen2_HVAR_pred,Number=./g' | \
        sed 's/dbNSFP_LRT_score,Number=A/dbNSFP_LRT_score,Number=./g' | \
        sed 's/dbNSFP_FATHMM_score,Number=A/dbNSFP_FATHMM_score,Number=./g' | \
        sed 's/dbNSFP_Polyphen2_HDIV_score,Number=A/dbNSFP_Polyphen2_HDIV_score,Number=./g' | \
        sed 's/dbNSFP_Interpro_domain,Number=A/dbNSFP_Interpro_domain,Number=./g' | \
        sed 's/RNA_ALT_FREQ,Number=1,Type=Integer/RNA_ALT_FREQ,Number=1,Type=Float/g' | \
        sed 's/RNA_ALT_COUNT,Number=1,Type=Float/RNA_ALT_COUNT,Number=1,Type=Integer/g' | \
	sed 's|##reference=file:///home/tgenref/pecan/refgenome/hs37d5/hs37d5_plusRibo_plusOncoViruses_plusERCC.fa|##reference=file:///home/tgenref/pecan/bwa_index/hs37d5_plusRibo_plusOncoViruses_plusERCC.fa|g' > ${ALL}.FIX

	if [ $? -ne 0 ]
        then
		rm snpEFF_on_merged_allTranscripts_in_progress
                touch snpEFF_on_merged_allTranscripts_Failed
                echo "### ${ALL} failed at Step 4) FIX" >> snpEFF_on_merged_allTranscripts_Failed
                rm snpEff_genes.txt
                rm snpEff_summary.html
                rm ${ALL}.noEFF
                rm ${ALL}.ANN
                rm ${ALL}.dbNSFP
                rm ${ALL}.FIX
        else
		VALIDATOR="`${VCFVALIDATOR} ${ALL}.FIX | wc -l`"
                if [ ${VALIDATOR} = "0" ]
                then
			rm snpEFF_on_merged_allTranscripts_in_progress
                	touch snpEFF_on_merged_allTranscripts_Complete
                	rm snpEff_genes.txt
                	rm snpEff_summary.html
                	rm ${ALL}.noEFF
                	rm ${ALL}.ANN
                	rm ${ALL}.dbNSFP
                	mv ${ALL}.FIX ${ALL}
		else
			rm snpEFF_on_merged_allTranscripts_in_progress
                	touch snpEFF_on_merged_allTranscripts_Failed
                	echo "### ${ALL} failed at Step 5) Validation" >> snpEFF_on_merged_allTranscripts_Failed
                	rm snpEff_genes.txt
                	rm snpEff_summary.html
                	rm ${ALL}.noEFF
                	rm ${ALL}.ANN
                	rm ${ALL}.dbNSFP
                	rm ${ALL}.FIX
		fi        
	fi
fi



####################################
#				   #
# Update Canonical Transcripts VCF #
#				   #
####################################

# Step 1) Remove EFF and dbNSFP annotations

java -jar ${SNPSIFT} rmInfo ${CANON} INDEL LOF NMD EFF dbNSFP_GERP++_RS dbNSFP_LRT_score dbNSFP_MutationAssessor_score dbNSFP_FATHMM_score dbNSFP_GERP++_NR dbNSFP_Polyphen2_HDIV_score dbNSFP_Polyphen2_HVAR_pred dbNSFP_SIFT_score dbNSFP_Polyphen2_HVAR_score dbNSFP_MutationTaster_score dbNSFP_Interpro_domain | \
        grep -v "##INFO=<ID=dbNSFP_" | \
        grep -v "##SnpSiftVersion" | \
        grep -v "##SnpEffVersion" | \
        grep -v "##SnpEffCmd" | \
	grep -v "##INFO=<ID=INDEL" | \
        grep -v "##INFO=<ID=EFF" | \
        grep -v "##INFO=<ID=LOF" | \
        grep -v "##INFO=<ID=NMD" | \
        grep -v "##INFO=<ID=SEURAT_IC," | \
        grep -v "##INFO=<ID=SEURAT_IHP," | \
        grep -v "##INFO=<ID=SEURAT_NT," | \
        grep -v "##INFO=<ID=SEURAT_OVERLAP," | \
        grep -v "##INFO=<ID=SEURAT_QSI," | \
        grep -v "##INFO=<ID=SEURAT_QSI_NT," | \
        grep -v "##INFO=<ID=SEURAT_RC," | \
        grep -v "##INFO=<ID=SEURAT_RU," | \
        grep -v "##INFO=<ID=SEURAT_SGT," | \
        grep -v "##INFO=<ID=SEURAT_SOMATIC," | \
        grep -v "##INFO=<ID=SEURAT_TQSI," | \
        grep -v "##INFO=<ID=SEURAT_TQSI_NT," | \
	grep -v "##FILTER=<ID=REJECT" | \
	sed 's|##FORMAT=<ID=MUTECT_AD,Number=.,Type=Integer,Description="Not provided in original VCF header">|##FORMAT=<ID=MUTECT_AD,Number=.,Type=Integer,Description="Allelic depths for the ref and alt alleles in the order listed">|g' | \
        sed 's|##FORMAT=<ID=MUTECT_BQ,Number=A,Type=Float,Description="Not provided in original VCF header">|##FORMAT=<ID=MUTECT_BQ,Number=A,Type=Float,Description="Average base quality for reads supporting alleles">|g' | \
        sed 's|##FORMAT=<ID=MUTECT_DP,Number=1,Type=Integer,Description="Not provided in original VCF header">|##FORMAT=<ID=MUTECT_DP,Number=1,Type=Integer,Description="Approximate read depth (reads with MQ=255 or with bad mates are filtered)">|g' | \
        sed 's|##FORMAT=<ID=MUTECT_FA,Number=A,Type=Float,Description="Not provided in original VCF header">|##FORMAT=<ID=MUTECT_FA,Number=A,Type=Float,Description="Allele fraction of the alternate allele with regard to reference">|g' | \
        sed 's|##FORMAT=<ID=MUTECT_GQ,Number=1,Type=Integer,Description="Not provided in original VCF header">|##FORMAT=<ID=MUTECT_GQ,Number=1,Type=Integer,Description="Genotype Quality">|g' | \
        sed 's|##FORMAT=<ID=MUTECT_GT,Number=1,Type=String,Description="Not provided in original VCF header">|##FORMAT=<ID=MUTECT_GT,Number=1,Type=String,Description="Genotype">|g' | \
        sed 's|##FORMAT=<ID=MUTECT_PL,Number=G,Type=Integer,Description="Not provided in original VCF header">|##FORMAT=<ID=MUTECT_PL,Number=G,Type=Integer,Description="Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification">|g' | \
        sed 's|##FORMAT=<ID=MUTECT_SS,Number=1,Type=Integer,Description="Not provided in original VCF header">|##FORMAT=<ID=MUTECT_SS,Number=1,Type=Integer,Description="Variant status relative to non-adjacent Normal,0=wildtype,1=germline,2=somatic,3=LOH,4=post-transcriptional modification,5=unknown">|g' | \
        sed 's|##INFO=<ID=MUTECT_DB,Number=0,Type=Flag,Description="Not provided in original VCF header">|##INFO=<ID=MUTECT_DB,Number=0,Type=Flag,Description="dbSNP Membership">|g' | \
        sed 's|##INFO=<ID=MUTECT_MQ0,Number=1,Type=Integer,Description="Not provided in original VCF header">|##INFO=<ID=MUTECT_MQ0,Number=1,Type=Integer,Description="Total Mapping Quality Zero Reads">|g' | \
        sed 's|##INFO=<ID=MUTECT_SOMATIC,Number=0,Type=Flag,Description="Not provided in original VCF header">|##INFO=<ID=MUTECT_SOMATIC,Number=0,Type=Flag,Description="Somatic event">|g' | \
        sed 's|##INFO=<ID=MUTECT_VT,Number=1,Type=String,Description="Not provided in original VCF header">|##INFO=<ID=MUTECT_VT,Number=1,Type=String,Description="Variant type, can be SNP, INS or DEL">|g' > ${CANON}.noEFF

if [ $? -ne 0 ]
then
        rm snpEFF_on_merged_canonicalOnly_in_progress
        touch snpEFF_on_merged_canonicalOnly_Failed
        rm ${CANON}.noEFF
        echo "### ${CANON} failed at Step 1) rmInfo" >> snpEFF_on_merged_canonicalOnly_Failed
fi

# Step 2) Add SnpEff ANN annotations (default with version 4.2)

java -Xmx14g -jar ${SNPEFF} -canon -c ${SNPEFFPATH}/snpEff.config -lof GRCh37.74 ${CANON}.noEFF | \
	java -jar ${SNPSIFT} varType - | \
        grep -v "##INFO=<ID=HOM,Number=A,Type=Flag,Description=\"Variant is homozygous\">" | \
        grep -v "##INFO=<ID=HET,Number=A,Type=Flag,Description=\"Variant is heterozygous\">" > ${CANON}.ANN

if [ $? -ne 0 ]
then
        rm snpEFF_on_merged_canonicalOnly_in_progress
        touch snpEFF_on_merged_canonicalOnly_Failed
        rm ${CANON}.ANN
        rm ${CANON}.noEFF
        echo "### ${CANON} failed at Step 2) SnpEff" >> snpEFF_on_merged_canonicalOnly_Failed
fi

# Step 3) Add dbNSFP annotations

java -jar ${SNPSIFT} dbnsfp \
        -v \
        -a \
        -db ${DBNSFP} \
        -f CADD_raw,CADD_raw_rankscore,CADD_phred,Interpro_domain,Polyphen2_HVAR_pred,GERP++_NR,GERP++_RS,LRT_score,MutationTaster_score,MutationAssessor_score,FATHMM_score,Polyphen2_HVAR_score,SIFT_score,Polyphen2_HDIV_score,MetaSVM_score,MetaSVM_rankscore,MetaSVM_pred,MetaLR_score,MetaLR_rankscore,MetaLR_pred,Reliability_index \
        ${CANON}.ANN > ${CANON}.dbNSFP

if [ $? -ne 0 ]
then
        rm snpEFF_on_merged_canonicalOnly_in_progress
        touch snpEFF_on_merged_canonicalOnly_Failed
        echo "### ${CANON} failed at Step 3) dbNSFP" >> snpEFF_on_merged_canonicalOnly_Failed
        rm snpEff_genes.txt
        rm snpEff_summary.html
        rm ${CANON}.noEFF
        rm ${CANON}.ANN
        rm ${CANON}.dbNSFP
fi

# Step 4) Fix header lines that are missing, have spaces, dbNSFP errors, RNA_ALT_FREQ float problem.

if [ ${CRNA} = "No"  ]
then
        ${BCFTOOLS} annotate -h ${HEADER} ${CANON}.dbNSFP | \
        sed 's/'\'' '\"'/'\'\"'/g' | \
	sed 's|##FILTER=<ID=PASS>|##FILTER=<ID=PASS,Description="Accept as a confident somatic mutation">|g' | \
	sed 's|##FILTER=<ID=PASS,Description="All filters passed">|##FILTER=<ID=PASS,Description="Accept as a confident somatic mutation">|g' | \
        sed 's/dbNSFP_CADD_raw,Number=A/dbNSFP_CADD_raw,Number=./g' | \
        sed 's/dbNSFP_CADD_raw_rankscore,Number=A/dbNSFP_CADD_raw_rankscore,Number=./g' | \
        sed 's/dbNSFP_CADD_phred,Number=A/dbNSFP_CADD_phred,Number=./g' | \
        sed 's/dbNSFP_Polyphen2_HVAR_pred,Number=A/dbNSFP_Polyphen2_HVAR_pred,Number=./g' | \
        sed 's/dbNSFP_GERP___NR,Number=A/dbNSFP_GERP___NR,Number=./g' | \
        sed 's/dbNSFP_GERP___RS,Number=A/dbNSFP_GERP___RS,Number=./g' | \
        sed 's/dbNSFP_LRT_score,Number=A/dbNSFP_LRT_score,Number=./g' | \
        sed 's/dbNSFP_MutationTaster_score,Number=A/dbNSFP_MutationTaster_score,Number=./g' | \
        sed 's/dbNSFP_MutationAssessor_score,Number=A/dbNSFP_MutationAssessor_score,Number=./g' | \
        sed 's/dbNSFP_MetaLR_rankscore,Number=A/dbNSFP_MetaLR_rankscore,Number=./g' | \
        sed 's/dbNSFP_MetaSVM_rankscore,Number=A/dbNSFP_MetaSVM_rankscore,Number=./g' | \
        sed 's/dbNSFP_MetaSVM_pred,Number=A/dbNSFP_MetaSVM_pred,Number=./g' | \
        sed 's/dbNSFP_MetaSVM_score,Number=A/dbNSFP_MetaSVM_score,Number=./g' | \
        sed 's/dbNSFP_MetaLR_pred,Number=A/dbNSFP_MetaLR_pred,Number=./g' | \
        sed 's/dbNSFP_MetaLR_score,Number=A/dbNSFP_MetaLR_score,Number=./g' | \
        sed 's/dbNSFP_Reliability_index,Number=A/dbNSFP_Reliability_index,Number=./g' | \
        sed 's/dbNSFP_Polyphen2_HVAR_score,Number=A/dbNSFP_Polyphen2_HVAR_score,Number=./g' | \
        sed 's/dbNSFP_SIFT_score,Number=A/dbNSFP_SIFT_score,Number=./g' | \
        sed 's/dbNSFP_Polyphen2_HVAR_pred,Number=A/dbNSFP_Polyphen2_HVAR_pred,Number=./g' | \
        sed 's/dbNSFP_LRT_score,Number=A/dbNSFP_LRT_score,Number=./g' | \
        sed 's/dbNSFP_FATHMM_score,Number=A/dbNSFP_FATHMM_score,Number=./g' | \
        sed 's/dbNSFP_Polyphen2_HDIV_score,Number=A/dbNSFP_Polyphen2_HDIV_score,Number=./g' | \
        sed 's/dbNSFP_Interpro_domain,Number=A/dbNSFP_Interpro_domain,Number=./g' | \
        sed 's/RNA_ALT_FREQ,Number=1,Type=Integer/RNA_ALT_FREQ,Number=1,Type=Float/g' | \
        sed 's/RNA_ALT_COUNT,Number=1,Type=Float/RNA_ALT_COUNT,Number=1,Type=Integer/g' | \
	sed 's|##reference=file:///home/tgenref/pecan/refgenome/hs37d5/hs37d5_plusRibo_plusOncoViruses_plusERCC.fa|##reference=file:///home/tgenref/pecan/bwa_index/hs37d5_plusRibo_plusOncoViruses_plusERCC.fa|g' > ${CANON}.FIX

        if [ $? -ne 0 ]
        then
                rm snpEFF_on_merged_canonicalOnly_in_progress
                touch snpEFF_on_merged_canonicalOnly_Failed
                echo "### ${CANON} failed at Step 4) FIX" >> snpEFF_on_merged_canonicalOnly_Failed
                rm snpEff_genes.txt
                rm snpEff_summary.html
                rm ${CANON}.noEFF
                rm ${CANON}.ANN
                rm ${CANON}.dbNSFP
                rm ${CANON}.FIX
        else
		# Step 5) Validation
		
		VALIDATOR="`${VCFVALIDATOR} ${CANON}.FIX | wc -l`"
                if [ ${VALIDATOR} = "0" ]
                then
                        rm snpEFF_on_merged_canonicalOnly_in_progress
                        touch snpEFF_on_merged_canonicalOnly_Complete
                        rm snpEff_genes.txt
                        rm snpEff_summary.html
                        rm ${CANON}.noEFF
                        rm ${CANON}.ANN
                        rm ${CANON}.dbNSFP
                        mv ${CANON}.FIX ${CANON}
                else
                        rm snpEFF_on_merged_canonicalOnly_in_progress
                        touch snpEFF_on_merged_canonicalOnly_Failed
                        echo "### ${CANON} failed at Step 5) Validation" >> snpEFF_on_merged_canonicalOnly_Failed
                        rm snpEff_genes.txt
                        rm snpEff_summary.html
                        rm ${CANON}.noEFF
                        rm ${CANON}.ANN
                        rm ${CANON}.dbNSFP
                        rm ${CANON}.FIX
                fi 
       fi
else
	sed 's/'\'' '\"'/'\'\"'/g' ${CANON}.dbNSFP | \
	sed 's|##FILTER=<ID=PASS>|##FILTER=<ID=PASS,Description="Accept as a confident somatic mutation">|g' | \
        sed 's/dbNSFP_CADD_raw,Number=A/dbNSFP_CADD_raw,Number=./g' | \
        sed 's/dbNSFP_CADD_raw_rankscore,Number=A/dbNSFP_CADD_raw_rankscore,Number=./g' | \
        sed 's/dbNSFP_CADD_phred,Number=A/dbNSFP_CADD_phred,Number=./g' | \
        sed 's/dbNSFP_Polyphen2_HVAR_pred,Number=A/dbNSFP_Polyphen2_HVAR_pred,Number=./g' | \
        sed 's/dbNSFP_GERP___NR,Number=A/dbNSFP_GERP___NR,Number=./g' | \
        sed 's/dbNSFP_GERP___RS,Number=A/dbNSFP_GERP___RS,Number=./g' | \
        sed 's/dbNSFP_LRT_score,Number=A/dbNSFP_LRT_score,Number=./g' | \
        sed 's/dbNSFP_MutationTaster_score,Number=A/dbNSFP_MutationTaster_score,Number=./g' | \
        sed 's/dbNSFP_MutationAssessor_score,Number=A/dbNSFP_MutationAssessor_score,Number=./g' | \
        sed 's/dbNSFP_MetaLR_rankscore,Number=A/dbNSFP_MetaLR_rankscore,Number=./g' | \
        sed 's/dbNSFP_MetaSVM_rankscore,Number=A/dbNSFP_MetaSVM_rankscore,Number=./g' | \
        sed 's/dbNSFP_MetaSVM_pred,Number=A/dbNSFP_MetaSVM_pred,Number=./g' | \
        sed 's/dbNSFP_MetaSVM_score,Number=A/dbNSFP_MetaSVM_score,Number=./g' | \
        sed 's/dbNSFP_MetaLR_pred,Number=A/dbNSFP_MetaLR_pred,Number=./g' | \
        sed 's/dbNSFP_MetaLR_score,Number=A/dbNSFP_MetaLR_score,Number=./g' | \
        sed 's/dbNSFP_Reliability_index,Number=A/dbNSFP_Reliability_index,Number=./g' | \
        sed 's/dbNSFP_Polyphen2_HVAR_score,Number=A/dbNSFP_Polyphen2_HVAR_score,Number=./g' | \
        sed 's/dbNSFP_SIFT_score,Number=A/dbNSFP_SIFT_score,Number=./g' | \
        sed 's/dbNSFP_Polyphen2_HVAR_pred,Number=A/dbNSFP_Polyphen2_HVAR_pred,Number=./g' | \
        sed 's/dbNSFP_LRT_score,Number=A/dbNSFP_LRT_score,Number=./g' | \
        sed 's/dbNSFP_FATHMM_score,Number=A/dbNSFP_FATHMM_score,Number=./g' | \
        sed 's/dbNSFP_Polyphen2_HDIV_score,Number=A/dbNSFP_Polyphen2_HDIV_score,Number=./g' | \
        sed 's/dbNSFP_Interpro_domain,Number=A/dbNSFP_Interpro_domain,Number=./g' | \
        sed 's/RNA_ALT_FREQ,Number=1,Type=Integer/RNA_ALT_FREQ,Number=1,Type=Float/g' | \
        sed 's/RNA_ALT_COUNT,Number=1,Type=Float/RNA_ALT_COUNT,Number=1,Type=Integer/g' | \
	sed 's|##reference=file:///home/tgenref/pecan/refgenome/hs37d5/hs37d5_plusRibo_plusOncoViruses_plusERCC.fa|##reference=file:///home/tgenref/pecan/bwa_index/hs37d5_plusRibo_plusOncoViruses_plusERCC.fa|g' > ${CANON}.FIX

        if [ $? -ne 0 ]
        then
                rm snpEFF_on_merged_canonicalOnly_in_progress
                touch snpEFF_on_merged_canonicalOnly_Failed
                echo "### ${CANON} failed at Step 4) FIX" >> snpEFF_on_merged_canonicalOnly_Failed
                rm snpEff_genes.txt
                rm snpEff_summary.html
                rm ${CANON}.noEFF
                rm ${CANON}.ANN
                rm ${CANON}.dbNSFP
                rm ${CANON}.FIX
        else
		# Step 5) Validation
	
		VALIDATOR="`${VCFVALIDATOR} ${CANON}.FIX | wc -l`"
		if [ ${VALIDATOR} = "0" ]
		then
                	rm snpEFF_on_merged_canonicalOnly_in_progress
                	touch snpEFF_on_merged_canonicalOnly_Complete
                	rm snpEff_genes.txt
                	rm snpEff_summary.html
                	rm ${CANON}.noEFF
                	rm ${CANON}.ANN
                	rm ${CANON}.dbNSFP
                	mv ${CANON}.FIX ${CANON}
		else
			rm snpEFF_on_merged_canonicalOnly_in_progress
                	touch snpEFF_on_merged_canonicalOnly_Failed
                	echo "### ${CANON} failed at Step 5) Validation" >> snpEFF_on_merged_canonicalOnly_Failed
                	rm snpEff_genes.txt
                	rm snpEff_summary.html
                	rm ${CANON}.noEFF
                	rm ${CANON}.ANN
                	rm ${CANON}.dbNSFP
                	rm ${CANON}.FIX
		fi
        fi
fi

if [ ! -f snpEFF_on_merged_canonicalOnly_Failed ] && [ ! -f snpEFF_on_merged_allTranscripts_Failed ]
then
	ssh ${USER}@172.19.1.224 "rsync ${DIR}/${CANON} ${STARTDIR}/${PATIENT_NAME}/vcfMerger/${EXOMEPAIR}/"
	
	if [ $? = 0 ]
	then
		ssh ${USER}@172.19.1.224 "rsync ${DIR}/${ALL} ${STARTDIR}/${PATIENT_NAME}/vcfMerger/${EXOMEPAIR}/"

		if [ $? = 0 ]
		then
			ssh ${USER}@172.19.1.224 "rm ${STARTDIR}/${PATIENT_NAME}/vcfMerger/${EXOMEPAIR}/SnpEFF_ANN_In_Progress ; \
				touch ${STARTDIR}/${PATIENT_NAME}/vcfMerger/${EXOMEPAIR}/SnpEFF_ANN_Complete ; \
				COMPLETE=0 ; \
				for XPAIR in \`echo ${EXOMEPAIRS} | sed 's/@/\n/g'\`
				do 
					if [ ! -f ${STARTDIR}/${PATIENT_NAME}/vcfMerger/\${XPAIR}/SnpEFF_ANN_Complete ] 
					then 
						COMPLETE=1  
					fi  
				done ; \
				if [ \${COMPLETE} = 0 ]
				then
					rm ${STARTDIR}/${PATIENT_NAME}/SnpEFF_ANN_In_Progress
					touch ${STARTDIR}/${PATIENT_NAME}/SnpEFF_ANN_Complete
				fi"
		else
			ssh ${USER}@172.19.1.224 "rm ${STARTDIR}/${PATIENT_NAME}/vcfMerger/${EXOMEPAIR}/SnpEFF_ANN_In_Progress ; \
				echo Failed to rsync the allTranscripts merged vcf back to isilon for ${EXOMEPAIR} >> ${STARTDIR}/${PATIENT_NAME}/vcfMerger/${EXOMEPAIR}/SnpEFF_ANN_Fail ; \
				echo Failed to rsync the allTranscripts merged vcf back to isilon for ${EXOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org "	
		fi
	else
		ssh ${USER}@172.19.1.224 "rm ${STARTDIR}/${PATIENT_NAME}/vcfMerger/${EXOMEPAIR}/SnpEFF_ANN_In_Progress ; \
                	echo Failed to rsync the canonical merged vcf back to isilon for ${EXOMEPAIR} >> ${STARTDIR}/${PATIENT_NAME}/vcfMerger/${EXOMEPAIR}/SnpEFF_ANN_Fail ; \
                        echo Failed to rsync the canonical merged vcf back to isilon for ${EXOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org "
	
	fi
else
	ssh ${USER}@172.19.1.224 "rm ${STARTDIR}/${PATIENT_NAME}/vcfMerger/${EXOMEPAIR}/SnpEFF_ANN_In_Progress ; \
        	echo SnpEFF failed somewere in the process for ${EXOMEPAIR} >> ${STARTDIR}/${PATIENT_NAME}/vcfMerger/${EXOMEPAIR}/SnpEFF_ANN_Fail ; \
                echo SnpEFF failed somewere in the process for ${EXOMEPAIR} | mailx -s "Post Medusa Processing Failed" ${USER}@tgen.org "

fi




